Add documentation around pipefail to address issue 1459

aeijdenberg opened the pull request:
Signed-off-by: Adam Eijdenberg <adam.eijdenberg@gmail.com>

### Proposed changes

Add documentation around `pipefail` to address issue 1459.

### Related issues (optional)

Fixes #1459.

mstanleyjones commented on engine/userguide/eng-image/dockerfile_best-practices.md:
```diff
@@ -283,6 +283,19 @@ keep the image size down. Since the `RUN` statement starts with
 
+#### Using pipes
+
+Where your `RUN` command depends on pipes, for example:
```
Not all of our users are UNIX experts. Let's give them some context, and also not start with a sentence fragment:

Some `RUN` commands depend on the ability to pipe the output of one command into another, using the pipe character (`|`), as in the following example:

mstanleyjones commented on engine/userguide/eng-image/dockerfile_best-practices.md:
```diff
@@ -283,6 +283,19 @@ keep the image size down. Since the `RUN` statement starts with
+Where your `RUN` command depends on pipes, for example:
+
+
+	RUN wget -O - https://some.site | wc -l > /number
```
Please use the triple-tick fencing syntax, with `Dockerfile` hint, instead of the 4-space indentation.

mstanleyjones commented on engine/userguide/eng-image/dockerfile_best-practices.md:
```diff
@@ -283,6 +283,19 @@ keep the image size down. Since the `RUN` statement starts with
+
+	RUN wget -O - https://some.site | wc -l > /number
+
+Remember that Docker executes this with `/bin/sh -c` which considers the command successful if the last operation in the pipe is successful. In our example above this build step will succeed and result in a new image produced so long as `wc -l` succeeds, even if the `wget` fails.
```
s/Remember that Docker executes this...successful./Docker executes these commands using the `/bin/sh -c` interpreter, which only evaluates the exit code of the last operation in the pipe to determine success.

s/In our example above/In the example above

s/will succeed and result/succeeds and produces a new image

s/the `wc -l` succeeds/the `wc -l` command succeeds (and the same for `wget`)

mstanleyjones commented on engine/userguide/eng-image/dockerfile_best-practices.md:
```diff
@@ -283,6 +283,19 @@ keep the image size down. Since the `RUN` statement starts with
+
+Remember that Docker executes this with `/bin/sh -c` which considers the command successful if the last operation in the pipe is successful. In our example above this build step will succeed and result in a new image produced so long as `wc -l` succeeds, even if the `wget` fails.
+
+To instead fail on an error at any stage in the pipe, prepend `set -o pipefail &&` to ensure that an unexpected error prevents the build from inadvertently succeeding, for example:
```
If you want the command to fail due to an error at any stage in the pipe,

aeijdenberg commented:
Hi @mstanleyjones , thanks for the suggested edits. I've changed as suggested and also added a note about shells that don't support `-o pipefail`.

If/when you're happy with the overall PR, I'm happy to rebase/squash the commits. Figured best keep them separate while we're still discussing changes.
